data-prepper-plugins/opensearch-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearch/worker/ScrollWorker.java

```diff
@@ -80,12 +89,16 @@ public void run() {
                 }
 
                 try {
-                    processIndex(indexPartition.get());
+                    final Pair<AcknowledgementSet, CompletableFuture<Boolean>> acknowledgementSet = createAcknowledgmentSet(
+                            acknowledgementSetManager,
+                            openSearchSourceConfiguration,
+                            sourceCoordinator,
+                            indexPartition.get());
 
-                    sourceCoordinator.closePartition(
-                            indexPartition.get().getPartitionKey(),
-                            openSearchSourceConfiguration.getSchedulingParameterConfiguration().getRate(),
-                            openSearchSourceConfiguration.getSchedulingParameterConfiguration().getJobCount());
+                    processIndex(indexPartition.get(), acknowledgementSet.getLeft());
+
+                    completeIndexPartition(openSearchSourceConfiguration, acknowledgementSet.getLeft(), acknowledgementSet.getRight(),
+                            indexPartition.get(), sourceCoordinator);
```

Comment:
It appears there is some opportunity to consolidate code in the workers. This behavior should be common and doesn't need to vary between workers, right?

Perhaps there can be a higher-level abstraction over the workers that handles the common behaviors. Or perhaps use a common base class.

Maybe this is something for a follow-on PR.

Reply:
Yes I didn't want to do a full refactor here as that would be pretty large, but I at least didn't duplicate the new code in this PR